src/Google.Api.Gax.Grpc/BufferedClientWriter.cs

```diff
@@ -0,0 +1,285 @@
+﻿/*
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ * Use of this source code is governed by a BSD-style
+ * license that can be found in the LICENSE file or at
+ * https://developers.google.com/open-source/licenses/bsd
+ */
+using Grpc.Core;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Google.Api.Gax.Grpc
+{
+    // TODO: Check whether cancellation token associated with original RPC is honoured throughout
+    // TODO: Check whether CallOptions are honoured throughout
+    // TODO: Check that the "further calls to Write after failure" handling is appropriate. (And document it!)
```

Comment:
What happens if one of the tasks returned by WriteAsync (et al) is cancelled?
It looks like that's ignored at the moment - is this the desired behaviour?